tests/integrations/raftstore/test_replica_read.rs

```diff
@@ -280,3 +280,41 @@ fn test_replica_read_on_stale_peer() {
     // must be timeout
     assert!(resp1_ch.recv_timeout(Duration::from_micros(100)).is_err());
 }
+
+#[test]
+fn test_read_index_out_of_order() {
+    let mut cluster = new_node_cluster(0, 2);
+
+    // Use long election timeout and short lease.
+    configure_for_lease_read(&mut cluster, Some(1000), Some(10));
+    cluster.cfg.raft_store.raft_store_max_leader_lease =
+        ReadableDuration(Duration::from_millis(100));
+
+    let pd_client = Arc::clone(&cluster.pd_client);
+    pd_client.disable_default_operator();
+
+    let rid = cluster.run_conf_change();
+    pd_client.must_add_peer(rid, new_peer(2, 2));
+
+    cluster.must_put(b"k1", b"v1");
+    must_get_equal(&cluster.get_engine(2), b"k1", b"v1");
+
+    cluster.must_transfer_leader(1, new_peer(1, 1));
+
+    let filter = Box::new(
+        RegionPacketFilter::new(1, 1)
+            .direction(Direction::Recv)
+            .msg_type(MessageType::MsgHeartbeatResponse)
+            .when(Arc::new(AtomicBool::new(true))),
+    );
+    cluster.sim.wl().add_recv_filter(1, filter);
+
+    let r1 = cluster.get_region(b"k1");
+    let resp = async_read_on_peer(&mut cluster, new_peer(1, 1), r1.clone(), b"k1", true, true);
+    assert!(resp.recv_timeout(Duration::from_secs(2)).is_err());
```

Comment:
How does the case fail without the fix?

Reply:
Without the fix the case will panic when waiting at line 319.